CSS adjustments in post-processed HTML manual

wikku opened the pull request:
This is a couple of small CSS changes to make the manual more legible and consistent with the PDF and old HTML manuals. The commit messages say a bit more.

Screenshot sample: htmlman, current webman, my proposed changes to webman
![image](https://user-images.githubusercontent.com/35867657/128924556-f2727ceb-576e-4fd9-9ee0-0fbaa6c53dae.png)

wikku committed:
webman: Remove colored background for inline code

Looks incohesive in syntax descriptions, unlike in htmlman or pdf which
don't color the backgrounds.

wikku committed:
webman: Follow inline code styling of htmlman

The color #888 has far too low contrast, I chose the color used for
links but darker.

wikku committed:
webman: Italicize grammar symbol names

Chapter 9, The OCaml language:
"Non-terminal symbols are set in italic font"

wikku committed:
webman: Make display style blocks centered

wikku committed:
webman: Increase heading font-weight

pmetzger commented:
No comment on the code, but the results in the image do look nicer.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -280,10 +283,17 @@ blockquote.quote{
     &::before{@include diamond}
 }
+.center {
+	text-align: center;
```
Horizontally aligning text on the center seems like a mistake outside of tables?

wikku replied:
It is necessary to center the displayed syntax fragments, as in the screenshot. These rules are just copied from the old HTML manual CSS.

Octachron replied:
It seems to me that having some indenting would be better than centering texts.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -283,7 +283,6 @@ blockquote.quote{
 span.c003{
     color:#564233;
     font-family:"Fira Mono",courier;
-    background-color:#f3ece6;
```
What is the motivation behind the change?

wikku replied:
The PDF and old HTML manuals don't color the background of inline code. 

In cases where you have syntax fragments interleaved with eg. grammar rules, this kind of styling can make the whole thing look incoherent, especially when they are inside a paragraph. You can take look at "[Local modules](https://ocaml.org/manual/expr.html)". Maybe it would be nice if we could make the whole construct "let module module-name =  module-expr in  expr" have a colored background with mixed typewriter and italic font inside, but we can't do it now since it's not a single HTML element.

Octachron replied:
The incoherence is a good point, thanks for pointing that out.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -293,7 +293,13 @@ div.caml-example.toplevel div.caml-input::before{
+span.c003, span.c005 {
+	color: rgba(91, 33, 6, 0.87);
+}
+span.c002{
```
What is the motivation for the color change on c0003, c0005 ? The font change?

wikku replied:
The font change for `.c004` and `.c005` sets respectively the `Match_failure` and `let` in the screenshot in monospace font. I tried to follow the CSS rules from the old HTML manual. I took the font-family from other parts of the file, but I added the `monospace` generic fallback at the end, as is good practice. Should I fix that in other places of the file while I'm here? 

 The color `#888888` that was used for (some) monospace fragments has simply too low contrast on `#ffffff`, fails [accessibility guidelines](https://webaim.org/resources/contrastchecker/) (even more on darker backgrounds like in tables) and looks grayed out and unimportant (to me, identifiers and syntax fragments are pretty important and should not blend into the background).  

Instead I chose a safe, non-offending color by simply darkening the color currently used for links (and adding a bit of transparency, to make it work better on darker backgrounds like in tables). Maybe a later PR will pick something better, but I wouldn't want to discuss this choice now.

Octachron replied:
I agree that switching the font for inline code to monospace is the right thing to do (we should get ride of those hevea generated classes at some points).
 
Adding a `monospace` failback to the CSS sounds like a good idea (in a separate commit probably).

I agree that the previous color has too little contrast. Having a color that is close from the grammatical symbol color seems like a sensible idea.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -316,6 +316,9 @@ span.authors{
 }
+span.c011 {
+	font-style: italic;
+}
```
Aligning the graphical style with the description is great indeed.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -185,6 +185,9 @@ h3 code{
 }
+h2, h3, h4 {
+	font-weight: 500;
+}
```
Some higher-level description of the change would be great here too.

wikku replied:
This change is non-essential and purely subjective. I think it made it easier to skim for subsection headings and adds some visual variation to the whole thing, gives less of a "wall of text" feeling.

Octachron replied:
I see your point and that seems potential nicer to me. It might be worthwhile to get a third advice (maybe on a larger text fragment to see how multiple titles mesh with the rest of the text?).

wikku replied:
Actually, this change is not without precedent, as both the old htmlman and LaTeX use a slightly higher font weight for those headings.

wikku committed:
webman: Use SCSS variables for font-family

wikku committed:
webman: Set command-line arguments in monospace

wikku committed:
webman: Set (mathematical) variables in italics

wikku committed:
webman: Remove styling for the ordinal "th"

The class .009 is used only for superscript "th" and it looked odd with
the colored background. Now it follows the LaTeX styling.

wikku committed:
webman/api: Improve constructor color contrast

Now the .constructor class (used for data constructors and modules) uses
the same color as types.

wikku commented:
I looked around the manual some more and noticed more places with bad/missing styling. The only controversial change is probably changing the constructor/module color in the API reference to be the same color as types. The previous color was #a28867 and it had very low contrast, especially on backgrounds such as #efeae7. With this change modules/constructors and types are of course still distinguishable by the capitalized first letter.

wikku committed:
webman: Center tables

wikku committed:
webman: Make table headers bold

Rules copied from htmlman

wikku committed:
webman: Fix variables being bolded, not italicized

Rule based on htmlman CSS

wikku committed:
webman: Add spacing to tables for aligned exprs

Rule copied from htmlman CSS

wikku committed:
webman: Add padding to .cellpadding1

Rule copied from htmlman CSS

wikku committed:
webman: Styling for Unix/Windows specific blocks

Similar to htmlman.
It's necessary to be able to see when Unix/Windows specific information
ends.  These blocks have to be styled differently, it's mentioned in the
foreword.

wikku committed:
webman: Italicize command-line argument variables

Rule based on htmlman CSS

wikku committed:
webman: Center BNF syntax bars `|`

Rule based on htmlman CSS

wikku commented:
I think this PR should go in 4.13 and perhaps 4.12 bugfix (and I don't think I can add the milestone), as it fixes some readability issues due to missing styling: 

* missing monospace, making some syntax fragments look like normal prose
* missing styling to delimit Unix/Windows specific manual fragments
* missing styling for command-line options, making the reader guess which part is static and which is a variable
* grammar symbols not italic as they should be according to the manual

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -304,7 +304,7 @@ div.caml-example.toplevel div.caml-input::before{
 }
-span.c004, span.c005 {
+span.c004, span.c005, span.c007 {
 	font-family: $font-mono;
```
Using a monospace fonts for command-line arguments make sense.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -323,6 +323,9 @@ span.c009{
     border-radius:6px
 }
+span.c010 {
+	font-style: italic;
```
This is in line with the simple html manual and the pdf one. Good.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -318,11 +318,6 @@ span.c006{
-span.c009{
-    font-style:italic;
-    background-color:#f3ece6;
-    border-radius:6px
```
This style was supposed to be a duplicate of c010 (due to a change of the hevea-generated html). I agree that simpler is better here.

Octachron commented on manual/src/html_processing/scss/style.scss:
```diff
@@ -793,8 +793,7 @@
     .caml-input span.kw2 {
-	font-weight: 500;
-	color: #a28867;
+	color: #8d543c;
```
Reducing the number of colors seems fine, but what is the reasoning behind reducing the font weight?

wikku replied:
I may have had another reason I can't remember, but I don't see why it would be bolded in the first place and it's not bolded in htmlman. I think now only keywords like `let` and `while` are bolded, which I think is standard when displaying code.

It's also annoying that `.constructor` in the HTML applies to both data constructors and modules, having a different style for modules (maybe just bolded) would be nice.

Octachron commented on manual/src/html_processing/scss/manual.scss:
```diff
@@ -280,6 +280,14 @@ blockquote.quote{
+	margin-right: auto;
+}
+.display {
+	margin: 0 auto;
```
In the old manual css this styling was on `.dcenter` which seems a little bit clearer in term of intents.

wikku replied:
The classes `.display` and `.dcenter` seem to always coexist, so it doesn't really matter. I don't think it's clearer, they're just different approaches ("semantic" vs "non-semantic" CSS)

Octachron reviewed (approved):
After rereading the changes, they seem coherent and motivated. Synchronizing the rendering of grammars and tables between the two version of the html manual is a good idea. 
I have few minor remarks but I am ok with the current state.

Would you mind adding a change entry in the 4.13 manual section?

wikku committed:
Add Changes entry

wikku commented:
Thanks for the review!

Octachron commented:
Thanks for the PR! (cherry-picked on 4.13 as bc139d6 and 4.12 as 148566022d)

sanette commented:
thanks @wiktorkuchta for the nice work. I noticed a side-effect, in the TOC
`webman/manual/index.html`
the part titles become centered, which, to me, is a visual design fault.
Of course this is not super important, because it doesn't affect legibility too much.

sanette commented:
another problem I see when setting the color of identifiers closer to black (in order to "getting a better contrast")
is that in fact it reduces the contrast with the rest of the text. I mean, when you have

"bla bla bla bla `let` bat bla bla"

the `let` identifier is not very visible and could be mistaken for a "normal" word. This is why I was in favor of styling it in a different way, for instance with a slightly shaded background. This is exactly what github does in the "bla" text above, and in fact, this could be consistent with the tutorials in the `ocaml.org` site, see
https://ocaml.org/learn/tutorials/guidelines.html

wikku commented:
The part titles have `.center`, so it's mainly a fault of the HTML we have. I don't think it looks bad though.

It's true that the color is closer to black and harder to distinguish from normal text, but there's still an improvement because some terminal symbols used to be unstyled and impossible to distinguish from normal text. In most cases the monospace font is enough to see that it's a terminal symbol, in fact that's all that the PDF manual does (it doesn't color them differently).

The colored background would look good in places like 

> The `let` and `let rec` constructs bind names locally.

but not really in places like

> The expression `let module` *module-name* `=` *module-expr* `in` *expr* locally binds

We can't color the background of the whole thing with the HTML we have.

sanette commented:
I agree that the HTML is not optimal. This will change of course when we switch to `odoc`.

> in fact that's all that the PDF manual does (it doesn't color them differently).

I wouldn't try to copy the PDF version as a rule, because the target of a PDF version is not the same as the one of a web version.

> 
> The colored background would look good in places like
> 
> > The `let` and `let rec` constructs bind names locally.
> 
> but not really in places like
> 
> > The expression `let module` _module-name_ `=` _module-expr_ `in` _expr_ locally binds
> 
>
to me, even in your last example, I find it better with background (like github does) than just monospace font. (easier to spot the keywords/symbols) Of course it's a matter of taste
